cli_tools/common/utils/daisyutils/daisy_worker.go

```diff
@@ -74,35 +78,93 @@ func createResourceLabelerIfMissing(env EnvironmentSettings, hooks []interface{}
 }
 
 type defaultDaisyWorker struct {
-	wf     *daisy.Workflow
-	logger logging.Logger
-	env    EnvironmentSettings
-	hooks  []interface{}
+	workflowProvider WorkflowProvider
+	finished         *daisy.Workflow
+	logger           logging.Logger
+	env              EnvironmentSettings
+	hooks            []interface{}
+
+	cancel      chan string
+	cancelGuard sync.Once
 }
 
 // Run runs the daisy workflow with the supplied vars.
-func (w *defaultDaisyWorker) Run(vars map[string]string) error {
-	if err := (&ApplyAndValidateVars{w.env, vars}).PreRunHook(w.wf); err != nil {
-		return err
+func (w *defaultDaisyWorker) Run(vars map[string]string) (err error) {
+	var wf *daisy.Workflow
+	for attempts := 0; attempts <= 1; attempts++ {
+		if wf, err = w.workflowProvider(); err != nil {
+			break
+		}
+		if err = w.checkIfCancelled(wf); err != nil {
```

Comment:
Why would a workflow be cancelled after just being created?

Reply:
There's a race where a caller could call `DaisyWorker.Cancel` before calling `Run`. A typical example would be a timeout that occurs in the period between construction and invocation of the DaisyWorker.

Here's the cancel method:

https://github.com/GoogleCloudPlatform/compute-image-tools/blob/85f7e15dfd425f494df6c7353953d92cc925e6fb/cli_tools/common/utils/daisyutils/daisy_worker.go#L190